File: macos/Sources/WGFileSender/Core/SharedConfig.swift
import Foundation

/// A file this device is offering to a peer, looked up by fileId when serving `/pull`.
struct OutgoingSource: Sendable {
    let path: String
    let size: Int64
    let sha256: String
}

/// Thread-safe snapshot of config the networking layer reads from background queues.
/// AppState (main actor) writes; the listener reads. Kept separate so the network
/// path never has to hop to the main actor just to look up a peer or the download root.
final class SharedConfig: @unchecked Sendable {
    private let lock = NSLock()
    private var _identity: Identity
    private var _peers: [PeerDevice]
    private var _settings: Store.Settings
    private var _outgoing: [String: OutgoingSource] = [:]   // fileId -> source the peer can pull

    init(identity: Identity, peers: [PeerDevice], settings: Store.Settings) {
        _identity = identity
        _peers = peers
        _settings = settings
    }

    private func sync<T>(_ body: () -> T) -> T {
        lock.lock(); defer { lock.unlock() }
        return body()
    }

    var identity: Identity { sync { _identity } }
    func setIdentity(_ value: Identity) { sync { _identity = value } }

    var peers: [PeerDevice] { sync { _peers } }
    func setPeers(_ value: [PeerDevice]) { sync { _peers = value } }

    var settings: Store.Settings { sync { _settings } }
    func setSettings(_ value: Store.Settings) { sync { _settings = value } }

    /// Find a peer by the token it presents (matches the token THIS device issued).
    func peer(forToken token: String) -> PeerDevice? {
        sync { _peers.first { $0.tokenIn == token } }
    }

    func peer(forId id: String) -> PeerDevice? {
        sync { _peers.first { $0.peerId == id } }
    }

    // MARK: outgoing sources (for serving /pull)

    func outgoingSource(_ fileId: String) -> OutgoingSource? { sync { _outgoing[fileId] } }
    func addOutgoing(_ fileId: String, _ source: OutgoingSource) { sync { _outgoing[fileId] = source } }
    func removeOutgoing(_ fileId: String) { sync { _outgoing.removeValue(forKey: fileId) } }
}

File: android/app/src/main/java/com/jaewonbaek/wgfilesender/data/SharedConfig.kt
package com.jaewonbaek.wgfilesender.data

import com.jaewonbaek.wgfilesender.model.Identity
import com.jaewonbaek.wgfilesender.model.PeerDevice
import com.jaewonbaek.wgfilesender.model.Settings
import com.jaewonbaek.wgfilesender.ui.Lang
import java.util.concurrent.ConcurrentHashMap

/** A file this device is offering to a peer, looked up by fileId when serving /pull. */
data class OutgoingSource(val uri: String, val size: Long, val sha256: String)

/** Snapshot the networking layer reads from background threads. */
class SharedConfig(identity: Identity, peers: List<PeerDevice>, settings: Settings) {
    @Volatile var identity: Identity = identity
    @Volatile var peers: List<PeerDevice> = peers
    @Volatile var settings: Settings = settings
    @Volatile var language: Lang = Lang.EN

    private val outgoing = ConcurrentHashMap<String, OutgoingSource>()   // fileId -> source

    /** Peer presenting [token] matches the token THIS device issued (tokenIn). */
    fun peerByToken(token: String): PeerDevice? = peers.firstOrNull { it.tokenIn == token }
    fun peerById(id: String): PeerDevice? = peers.firstOrNull { it.peerId == id }

    fun outgoingSource(fileId: String): OutgoingSource? = outgoing[fileId]
    fun addOutgoing(fileId: String, src: OutgoingSource) { outgoing[fileId] = src }
    fun removeOutgoing(fileId: String) { outgoing.remove(fileId) }
}
